protocols/kad/src/behaviour.rs

```diff
@@ -2412,16 +2445,26 @@ pub enum KademliaEvent {
     // is made of multiple requests across multiple remote peers.
     InboundRequest { request: InboundRequest },
 
-    /// An outbound query has produced a result.
+    /// An outbound query has finished.
     OutboundQueryCompleted {
         /// The ID of the query that finished.
         id: QueryId,
-        /// The result of the query.
+        /// The optional final result of the query.
         result: QueryResult,
         /// Execution statistics from the query.
         stats: QueryStats,
     },
 
+    /// An outbound query has produced a result.
+    OutboundQueryProgressed {
```

Comment:
_Just a slight preference for now. Should probably put more thoughts into this._

How about the following:

- Remove `OutboundQueryProgressed`.
- Rename `OutboundQueryCompleted` to `OutboundQueryProgressed`.
- Add field `index: Index` to `OutboundQueryProgressed` (formerly `OutboundQueryCompleted`) where `enum Index { Continuing(usize), Final(usize) }` indicating the sequence number and whether it is the final update for the query.

Reply:
I don't have a strong preference of my solution over this, happy to change it, but lets get decide on a version before I go change it again 😅

Reply:
Unless anyone of the recently tagged people (or anyone else) raises any objections, I suggest going with the above.

Reply:
> I am unfamiliar with the internals of kademlia but do we know ahead of time, how many we will get?

At least not in the providers case

Reply:
> What does the number within `Final` mean? Would it work to include a boolean `completed` in the event and only have `count` otherwise?

Yep, that is fine as well. I want to provide two pieces of information. (1) how many of these progress updates for the same query have been emitted thus far and (2) whether this one is the final one, i.e. the query overall finished.

Reply:
@mxinden I have started implementing this, but running into challenges with the `QueryStats`. The current design assumes that these can be only extracted once the query is completed, but with your suggested change we would want to provide stats on every progress event. Thoughts?

Reply:
> but running into challenges with the `QueryStats`. The current design assumes that these can be only extracted once the query is completed

I don't see why `QueryStats` could not be returned on each `OutboundQueryProgressed`. The `QueryStats` of a `Query` are updated on each request, thus they are always up-to-date. Am I missing something @dignifiedquire?

I am also fine returning them on the final event only, e.g. embedding them as: `Index { Continuing(usize), Final(usize, QueryStats) }` (ideally more self describing than a tuple).

Reply:
@mxinden I started expanding this to `GetRecord` and am running into issues with this construction. There is a need to differentiate the type and information between the progress events and the last event, e.g. for `GetRecord` the last event will not contain an actual record, only the optional cache. 
Thoughts?

Reply:
How about instead returning the `cache_candidates` on each progression? I would expect this `BTreeMap` to be small and thus cloning to be cheap. Would you agree @dignifiedquire? Would that solve the issue?

Reply:
not entirely, as I would still need to make the record be optional, which is pretty confusing as a consumer

Reply:
Or we return an event for each contacted peer directly with an optional record.